File: ruby/palindrome-products/palindrome_products.rb
Palindrome = Struct.new(:value, :factors)

class Palindromes
  def initialize(conditions)
    @min = conditions.fetch(:min_factor) {1}
    @max = conditions.fetch(:max_factor)
    @range = [*@min..@max]
  end

  def generate
    @palindromes = {}
    @range.repeated_combination(2).map do |factors|
      product = factors.first * factors.last
      if is_palindrome?(product)
        palindrome = @palindromes[product] || Palindrome.new(product, [])
        palindrome.factors << factors.sort
        palindrome.factors.uniq!
        @palindromes[product] = palindrome
      end
    end
  end

  def is_palindrome?(int)
    int.to_s == int.to_s.reverse
  end

  def sort
    @palindromes.sort_by { |key, _palindrome| key }
  end

  def largest
    sort.last[1]
  end

  def smallest
    sort.first[1]
  end
end


# palindromes = Palindromes.new(max_factor: 9)
# p palindromes.generate

File: javascript/palindrome-products/palindrome-products.js
const isPalindrome = n => parseInt(n.toString().split('').reverse().join('')) === n;

const areSameArray = (arr1, arr2) => arr1.toString() === arr2.toString();

const generate = (params) => {
  const minFactor = params.minFactor || 1;
  const maxFactor = params.maxFactor;

  const palindromes = {};

  for (let i = minFactor; i <= maxFactor; i++) {
    for (let j = i; j <= maxFactor; j++) {
      let factors = [i, j].sort();
      let product = i * j;
      if (isPalindrome(product) && palindromes[product]) {
        if (!palindromes[product].some(areSameArray)) {
          palindromes[product].push(factors);
        }
      } else if (isPalindrome(product) && palindromes[product] === undefined) {
        palindromes[product] = [factors];
      }
    }
  }

  console.log(palindromes)


  // return {
  //   largest: {
  //     value: maxProduct,
  //     factors: data[maxProduct],
  //   },
  //   smallest: {
  //     value: minProduct,
  //     factors: data[minProduct],
  //   },
  // };
};

if (module) { module.exports = generate };

// console.log(isPalindrome(102));
// console.log(range(1, 5));

generate({ maxFactor: 9 });

